### PR TITLE
HTCONDOR-2134 job-cool-down

### PR DESCRIPTION
**Pull request by JaimeFrey**
<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [ ] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [ ] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [ ] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [ ] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [ ] Verify that the branch destination of the PR matches the target version of the ticket
- [ ] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [ ] Check for documentation, if needed  (documentation [build logs](https://readthedocs.org/projects/htcondor/builds/))
- [ ] Check for version history, if needed
- [ ] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [ ] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

**Review by johnkn — CHANGES_REQUESTED** [review]

File: src/condor_schedd.V6/schedd_stats.h [hunk 1]
```diff
@@ -72,6 +72,7 @@ typedef struct ScheddJobCounters {
    stats_entry_recent<int> JobsMissedDeferralTime;    // jobs that exited with JOB_MISSED_DEFERRAL_TIME
    stats_entry_recent<int> JobsShouldHold;            // jobs that exited with JOB_SHOULD_HOLD
    stats_entry_recent<int> JobsDebugLogError;         // jobs that exited with DPRINTF_ERROR
+   stats_entry_recent<int> JobsCoolDowns;             // jobs that entered cool-down
```
Comment (johnkn) [inline src/condor_schedd.V6/schedd_stats.h:75]: all of the other plural stats are histograms.

File: src/condor_shadow.V6.1/baseshadow.cpp [hunk 2]
```diff
@@ -346,30 +346,25 @@ int BaseShadow::cdToIwd() {
...
 		DC_Exit( reason );
 	}
 		//Attempt to write Job ad to epoch file
 		//If knob isn't set or there is no job ad the function will just log and return
-	writeJobEpochFile(getJobAd());
-		// if we are being called from the exception handler, return
-		// now to prevent infinite loop in case we call EXCEPT below.
-	if ( reason == JOB_EXCEPTION ) {
-		return;
-	}
```
Comment (johnkn) [inline src/condor_shadow.V6.1/baseshadow.cpp:363]: I don't see how the new code handles this case.

Comment (JaimeFrey) [reply src/condor_shadow.V6.1/baseshadow.cpp:363]: I didn't see a way for the exception handler (or any code path using JOB_EXCEPTION) to end up here.

Comment (johnkn) [reply src/condor_shadow.V6.1/baseshadow.cpp:363]: Fair enough.  that code is from 2002, so maybe its no longer needed.

**Review by johnkn — APPROVED** [review]
Looks good.

File: src/condor_schedd.V6/schedd_stats.h (hunk 1 above)
Comment (JaimeFrey) [reply src/condor_schedd.V6/schedd_stats.h]: Fixed

File: src/condor_schedd.V6/schedd.cpp [hunk 3]
```diff
@@ -12417,6 +12422,35 @@ Scheduler::jobExitCode( PROC_ID job_id, int exit_code )
 			srec->reconnect_done = true;
 		}
 	}
+	if (exit_code == JOB_SHOULD_REQUEUE || exit_code == JOB_NOT_STARTED) {
+		long long ival = 0;
+		classad::Value val;
+		ClassAd * job_ad = GetJobAd( job_id.cluster, job_id.proc );
```
[the remark this answers is not in the document]
Comment (JaimeFrey) [reply src/condor_schedd.V6/schedd.cpp]: Fixed

**Pull request merged**